pkg/volume/csi/csi_client_test.go

```diff
@@ -338,3 +388,100 @@ func TestClientNodeUnstageVolume(t *testing.T) {
 		}
 	}
 }
+
+func TestGetNodeInfoWithFallback(t *testing.T) {
+	var errRandom = errors.New("some other error")
+	var errUnimplemented = status.Error(codes.Unimplemented, "unknown method someRandomMethod")
+
+	testCases := map[string]struct {
+		nodeGetInfoErr          error
+		nodeGetIdErr            error
+		expectedErr             error
+		expectedGetNodeInfoCall bool
+		expectedGetNodeIdCall   bool
+	}{
+		"no error on neither method": {
+			nil,
+			nil,
+			nil,
+			true,
+			false,
+		},
+		"unimplemented NodeGetInfo": {
+			errUnimplemented,
+			nil,
+			nil,
+			true,
+			true,
+		},
+		"other error for NodeGetInfo": {
+			errRandom,
+			nil,
+			errRandom,
+			true,
+			false,
+		},
+		"unimplemented NodeGetInfo and error on NodeGetId": {
+			errUnimplemented,
+			errRandom,
+			errRandom,
+			true,
+			true,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			ctx := context.TODO()
+			ig := &nodeInfoGetter{
+				nodeGetInfoErr: tc.nodeGetInfoErr,
+				nodeGetIdErr:   tc.nodeGetIdErr,
+			}
+			_, _, _, err := getNodeInfoWithFallback(ctx, ig)
+			if err != tc.expectedErr {
+				t.Errorf("Expected error %v, got: %v", tc.expectedErr, err)
+			}
+			checkCall(t, ig.nodeGetInfoCallCount, tc.expectedGetNodeInfoCall, "NodeGetInfo")
+			checkCall(t, ig.nodeGetIdCallCount, tc.expectedGetNodeIdCall, "NodeGetId")
+		})
+	}
+}
+
+func checkCall(t *testing.T, callCount int, expectedCall bool, funcName string) {
+	t.Helper()
+
+	if expectedCall {
+		if callCount < 1 {
+			t.Errorf("Expected %s to be called, was not called", funcName)
+		}
+		return
+	}
+
+	if callCount > 0 {
+		t.Errorf("Expected %s not to be calle, was called %d times", funcName, callCount)
+	}
+}
+
+type nodeInfoGetter struct {
+	nodeGetIdErr   error
+	nodeGetInfoErr error
+
+	nodeGetIdCallCount   int
+	nodeGetInfoCallCount int
+}
+
+func (g *nodeInfoGetter) NodeGetId(ctx context.Context) (string, error) {
+	g.nodeGetIdCallCount += 1
+	if err := g.nodeGetIdErr; err != nil {
```

Comment:
nit: `if g.nodeGetIdErr != nil`

Reply:
We were setting this `err` var here in the initializer statement, so that we have a shorthand to the error in the first branch of that if. However we can rewrite that to not use the initialize and also use `g.nodeGetIdErr` in the branch if you prefer that.

Reply:
i see, either way is fine